fix(forms): don't prevent default behavior for forms with method="dialog"

crisbeto opened the pull request:
The forms `submit` event handlers have a `return false` to prevent form submissions from reloading the page, however this also prevents the browser behavior for forms with `method="dialog"`.

These changes add an exception since the `method="dialog"` doesn't refresh the page.

Fixes #47150.

crisbeto commented on packages/forms/src/directives/reactive_directives/form_group_directive.ts:
```diff
@@ -94,8 +94,9 @@ export class FormGroupDirective extends ControlContainer implements Form, OnChan
   constructor(
-      @Optional() @Self() @Inject(NG_VALIDATORS) private validators: (Validator|ValidatorFn)[],
-      @Optional() @Self() @Inject(NG_ASYNC_VALIDATORS) private asyncValidators:
+      private _elementRef: ElementRef,
```
I wanted to use `inject` here so the constructor signature doesn't change, but we have some tests where the directive is constructed manually.

AndrewKushnir commented on packages/forms/src/directives/ng_form.ts:
```diff
@@ -133,6 +133,7 @@ export class NgForm extends ControlContainer implements Form, AfterViewInit {
   @Input('ngFormOptions') options!: {updateOn?: FormHooks};
 
   constructor(
+      private _elementRef: ElementRef,
```
Can we handle the `method` as an input instead?

```
@Input() method: string;
```

crisbeto replied:
Sure, that's also an option but we'll need a host binding for it too.

AndrewKushnir replied:
It looks like we only read the value, why would we need a host binding for it?

crisbeto replied:
It's because by having an input matching method we would have to forward it back to the DOM so the browser can pick it up.

AndrewKushnir replied:
Right, we'd need to retain it in the DOM as well... In this case, injecting an ElementRef looks good 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

crisbeto commented:
I ended up with the `event.target` workaround that we talked about. These are previous approaches and why they didn't work:
* We can't inject the `ElementRef` using the constructor in order to figure out the `method`, because there are a bunch of internal tests that construct the `NgForm` and `FormNameDirective` classes manually.
* We can't use the `inject` function to inject the `ElementRef` due to the aforementioned issue where internal targets instantiate the directives manually which means that there won't be an injection context.
* We can't use an `Input` to get the `method` of the form element, because the `method` input also has to be forwarded back to the `method` attribute on the host in order for the form to work correctly. The directive selectors allow for them to be set on an `ng-container` which will throw an error if we try to bind to an attribute.

AndrewKushnir reviewed (approved):
👍

dylhunn reviewed (approved):
reviewed-for: fw-core, fw-forms

pkozlowski-opensource reviewed (approved):
LGTM

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:472638088:BASE:472638209:1662529037282:22335f6d).

crisbeto commented:
It looks like some targets failed internally because `onSubmit` was being invoked without an event object. I've pushed a change to work around it.

AndrewKushnir commented:
@crisbeto thanks for the fix 👍
Could you please also add a quick comment around these lines to provide more info on why `$event` and `$event.target` might be undefined?
```
return ($event?.target as HTMLFormElement | null)?.method === 'dialog';
```

crisbeto committed:
fix(forms): don't prevent default behavior for forms with method="dialog"

The forms `submit` event handlers have a `return false` to prevent form submissions from reloading the page, however this also prevents the browser behavior for forms with `method="dialog"`.

These changes add an exception since the `method="dialog"` doesn't refresh the page.

Fixes #47150.

crisbeto commented:
Commit has been added.

AndrewKushnir commented:
[Global Presubmit #2](https://test.corp.google.com/OCL:472638088:BASE:472890608:1662611888938:a4421ef5).

dylhunn commented:
This PR was merged into the repository by commit 4a13210ecd692aaace7f67fadf25741f3f0cc50b.
